Python/Product/PythonTools/PythonTools/Project/ConfigureTestFrameworkInfoBar.cs

```diff
@@ -32,85 +32,133 @@
+                }
 
+                acceptActionItems.Add(new InfoBarHyperlink(Strings.UnitTestEnableInfoBarAction, (Action)EnableUnitTestAction));
             } else {
                 return;
             }
 
-            ShowInfoBar(infoBarMessage, acceptActionItem);
+            ShowInfoBar(infoBarMessage, acceptActionItems);
         }
 
-        private void ShowInfoBar(string infoBarMessage, InfoBarHyperlink acceptActionItem) {
+
+        private void ShowInfoBar(string infoBarMessage, List<InfoBarHyperlink> acceptActionItems) {
             LogEvent(ConfigurePytestInfoBarActions.Prompt);
-            var messages = new List<IVsInfoBarTextSpan>();
-            var acceptAction = new List<InfoBarActionItem>();
 
-            messages.Add(new InfoBarTextSpan(infoBarMessage));
-            acceptAction.Add(acceptActionItem);
-            acceptAction.Add(new InfoBarHyperlink(Strings.PyTestIgnoreInfoBarAction, (Action)IgnoreAction));
+            acceptActionItems.Add(new InfoBarHyperlink(Strings.PythonTestFrameworkIgnoreInfoBarAction, (Action)IgnoreAction));
+            Create(new InfoBarModel(
+                new List<IVsInfoBarTextSpan> { new InfoBarTextSpan(infoBarMessage) },
+                acceptActionItems,
+                KnownMonikers.StatusInformation)
+            );
+        }
+
+        private async Task<bool> InstallPyTestAsync() {
+            var packageManagers = _infoBarData.InterpreterOptionsService.GetPackageManagers(_infoBarData.InterpreterFactory);
+
+            foreach (var packageManager in packageManagers) {
+                bool pytestInstalled = await packageManager.InstallAsync(
+                    new PackageSpec("pytest"),
+                    new VsPackageManagerUI(Site),
+                    CancellationToken.None
+                );
+
+                if (pytestInstalled) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private async Task<bool> IsPyTestInstalledAsync() {
+            var packageManagers = _infoBarData.InterpreterOptionsService.GetPackageManagers(_infoBarData.InterpreterFactory);
+            var getPackageManagersTask = packageManagers.Select(
+                packageManager => packageManager.GetInstalledPackagesAsync(CancellationToken.None)
+            ).ToList();
 
-            Create(new InfoBarModel(messages, acceptAction, KnownMonikers.StatusInformation));
+            var packagesList = await Task.WhenAll(getPackageManagersTask);
+            return packagesList.Any(packages => packages.Any(package =>
+                String.Equals(package.Name, nameof(TestFrameworkType.Pytest), StringComparison.OrdinalIgnoreCase))
+            );
+        }
+
+        private bool PythonTestFileFound() {
+            return PathUtils.EnumerateFiles(_infoBarData.RootDirectory, "*test*.py", true).Any();
```

Comment:
For projects, this will result in false positives if the test files are on disk but not in the project.
For workspaces, every file is considered "in project" except for the Python environment folder. Ex: User creates a virtual environment 'env' under their workspace root folder and then install packages that have 'test' files. That shouldn't trigger a prompt (on next project load).

Reply:
https://github.com/microsoft/PTVS/issues/5508